### PR TITLE
Add copy() method to graph classes

### PR DESCRIPTION
**Pull request by mtreinish**
This commit adds a new copy() method to the graph classes to return a
shallow copy (with shared references to weight/data payload objects) of
a graph object.

### CODE REVIEW

**Comment by mtreinish** [comment]
I was curious how this compared to doing a python shallow copy with `copy.copy()` and put together a quick benchmark:

```python
import statistics
import time
import copy

import retworkx


graph = retworkx.directed_gnm_random_graph(6000, 5000)

copy_times = []
for _ in range(100000):
    start = time.time()
    graph.copy()
    stop = time.time()
    copy_times.append(stop - start)

copy_copy_times = []
for _ in range(100000):
    start = time.time()
    copy.copy(graph)
    stop = time.time()
    copy_copy_times.append(stop - start)

print("mean graph.copy() time: %s sec" % statistics.mean(copy_times))
print("mean copy.copy(graph) time: %s sec" % statistics.mean(copy_copy_times))
```

which for me locally resulted in:

```
mean graph.copy() time: 9.36128568649292e-05 sec
mean copy.copy(graph) time: 0.0011064216804504395 sec
```

File: tests/digraph/test_copy.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import unittest
+
+import retworkx
+
+
+class TestDeepcopy(unittest.TestCase):
```
Comment (t-imamichi) [inline tests/digraph/test_copy.py:18]: Isn't it test for shallow copy?

**Review by t-imamichi — APPROVED** [review]
LGTM

File: tests/digraph/test_copy.py (hunk 1 above)
Comment (mtreinish) [reply tests/digraph/test_copy.py:18]: Oops, yeah good catch:

```suggestion
class TestCopy(unittest.TestCase):
```

File: tests/graph/test_copy.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import unittest
+
+import retworkx
+
+
+class TestDeepcopy(unittest.TestCase):
```
Comment (mtreinish) [inline tests/graph/test_copy.py:18]:
```suggestion
class TestCopy(unittest.TestCase):
```

**Pull request merged**